### ISSUE TITLE
rethink constants

### ISSUE THREAD

**Issue by mzuenni** [Kattis/problem-package-format#344]
Right now we replace constants at the following places:

1. problem statements
2. input and output validators
3. included code
4. example submissions
5. `testdata.yaml`

There are a few things that I want to mention.

1. "Replacement in example submissions and included code" What is the use case here and why do we differentiate between example submissions and other submissions? (Note that I don't want to argue that we should do replacement for all submissions, rather I think that we should not touch any code at all).

2. In most cases text replacement is just inconvenient. We already restrict validators to c++ and python in both cases there are nicer ways to automagically introduce constants. For example in c++ we could add `constexpr <std::string_view,long long,double> <constant name> = <constant value>` and something similar in python. In latex, we could also introduce the constants as macros.
I know that this makes constants more complicated to implement on the tooling side but IMO it is much more convenient for the user. (For markdown and `testdata.yaml` i have no better idea...)

**Comment by RagnarGrootKoerkamp** [comment]
Text replacement badly messes up editor highlighting and autoformatting and such. Wouldn't it be nicer in your example if we did a `#define MAX_N 1000` in the background? And similarly we could `\newcommand{\maxn}{1000}` for latex.

**Comment by evouga** [comment]
Is there concrete evidence that the proposed balanced-braces “badly” messes up anything in practice?

I think it is good to make it obvious which variables are externally defined problem constants and which are statically scoped, and to make it more or less impossible to accidentally shadow the problem constants

**Comment by mzuenni** [comment]
regarding the change in judge submissions: I am not sure if this really makes things much easier since you still need to change all the test data anyway (note that we don't define generators). And distinguishing between user submissions and jury submissions makes stuff more complicated than necessary. Further, we don't don't allow a constant to write solutions like "do random stuff until you are out of time" for which automatic updates would be useful.

regarding messing stuff up: my problem is more that this messes up standard build processes. For c++ I can't compile stuff anymore but if it's done with macros or as constant I can simply tell my compiler what macros exist. Same for latex. The curly bracket thing breaks latexmk rebuild whereas with macros this can not happen.

**Comment by Tagl** [comment]
Yeah I think defines and environment variables might be better approaches here.

**Comment by evouga** [comment]
For LaTeX, don't you already need a script that does preprocessing before being able to compile the problem statement? You need to wrap the problem statement in a `document` environment, define all of the problem-package-specific macros, etc.

I agree that it's annoying that C++ code cannot be compiled locally using e.g. `-D{{maxn}}=100000`, and that it may be quite difficult to convince my judges to use constants in the judge solutions for this reason.

So:
 - we could try to come up with some language-specific way to include the constants (note that this would need to include all valid input validator languages, including stuff like CheckTestData, not just Python, TeX, and C++).
 - we could remove judge submissions from the list of preprocessed files, if this is the main sticking point. I am OK with this, though I don't really see the harm---judge teams who can't be bothered to deal with problem constants during local compilation don't have to use them.

But we had long discussions already about these issues at Lund, and considered alternatives like `__maxN__` and `XXMAXNXX` and other schemes that could be implemented as macros, and the status quo was a compromise balancing factors such as simplicity, flexibility across multiple languages, readability, and not-breaking-syntax-highlighting-too-badly. I don't think the current proposal fits everybody's workflow perfectly, but I'm also hesitant to relitigate the feature from scratch, unless there is a **major** undiscovered problem with the proposed approach.

**Comment by mzuenni** [comment]
> For LaTeX, don't you already need a script that does preprocessing before being able to compile the problem statement?

you can make latexmk do all that. it just cant preprocess the tex file with text replacement... (especially not with live pdf generation)

>we could try to come up with some language-specific way to include the constants (note that this would need to include all valid input validator languages, including stuff like CheckTestData, not just Python, TeX, and C++).

we only allow ctd, viva, c++ and python as checkers

**Comment by evouga** [comment]
> we only allow ctd, viva, c++ and python as checkers

Yes, you are right. So we could specify different mechanisms for these languages (plus yaml and markdown) and remove support for constants in included code and submissions.

> you can make latexmk do all that. it just cant preprocess the tex file with text replacement... (especially not with live pdf generation)

And it cannot be configured to `sed` the input files? (Though this is a moot point if you have a nice solution along the lines of the above).

**Comment by mzuenni** [comment]
>>you can make latexmk do all that. it just cant preprocess the tex file with text replacement... (especially not with live pdf generation)
>
>And it cannot be configured to sed the input files? (Though this is a moot point if you have a nice solution along the lines of the above).

I see no good way to do that. but if we add a mechanism for c++ etc. then one for latex is also fine i.e. if you have the constant `blub` we define the latex macro `\blub` or `\constant_blub` or something like this.

**Comment by niemela** [comment]
> 1. why do we differentiate between example submissions and other submissions?

"example submissions" are the only kind of submissions in the problem packages.

> Expanding constants in _contestant_ submissions seems to me a bad idea; [...]

Yes, definitely. I don't think anybody has suggested or want that.

> regarding the change in judge submissions: I am not sure if this really makes things much easier since you still need to change all the test data anyway (note that we don't define generators). 

I agree that constants are significantly less needed for (judge) submissions, simply because the agreement between submissions and statement/data/validators are checked. If there is a change of a limit that breaks a submission, then installation will fail.

The reason why we wanted something like constants to begin with is that we verify that submissions agree with data/validators, but we have to trust problem authors to make sure that the *statement* also agrees. In fact, the most common error with problem packages is exactly this.

> And distinguishing between user submissions and jury submissions makes stuff more complicated than necessary. 

I don't understand what this means? There are no "user submissions" in a problem package. Clearly (?) constants are not meant to apply outside of the package, i.e. they are absolutely not meant to be applied to submissions sent to some system that has installed the package.

> Further, we don't don't allow a constant to write solutions like "do random stuff until you are out of time" for which automatic updates would be useful.

I can't quite parse this. Is "contsant" supposed to be "contestant"? If so, why would constants help with writing "do stuff until time runs out" style submissions?

> > we only allow ctd, viva, c++ and python as checkers
> 
> Yes, you are right. So we could specify different mechanisms for these languages (plus yaml and markdown) and remove support for constants in included code and submissions.

A minor point is that this could make it harder to add additional languages in the future. That said, we could always say that constants are not avaialble for some languages (and then we would be not worse off than if we didn't define constants).

**Comment by mzuenni** [comment]
> > 1. why do we differentiate between example submissions and other submissions?
> 
> "example submissions" are the only kind of submissions in the problem packages.

yes but in the end the problem gets uploaded to a a system that handles both the example submissions and participants submissions. Now it has to distinguish them (which will likely also make it impossible to upload the example submission alone). 

> > Further, we don't don't allow a constant to write solutions like "do random stuff until you are out of time" for which automatic updates would be useful.
> 
> I can't quite parse this. Is "contsant" supposed to be "contestant"? If so, why would constants help with writing "do stuff until time runs out" style submissions?

Regarding use cases for constants in submissions: I think inserting something like the time-limit into a submission would be more useful than inserting constants but we already decided that that won't be possible.

 Overall i am just in favor of not having constants in submissions and use the for their main purpose, to make checkers and statements consistent.



> > > we only allow ctd, viva, c++ and python as checkers
> > 
> > 
> > Yes, you are right. So we could specify different mechanisms for these languages (plus yaml and markdown) and remove support for constants in included code and submissions.
> 
> A minor point is that this could make it harder to add additional languages in the future. That said, we could always say that constants are not avaialble for some languages (and then we would be not worse off than if we didn't define constants).

that is true, but I would prefer to change the constants like this:

- If its c++ do _X_ (for example macros)
- If its python do _Y_ (for example insert a variable with that value)
- If its tex do _Z_ (for example introduce a macro)
- Else, replace `{{constant}}` with its value (we don't (yer?) have better ideas for ctd checkers, viva checkers, markdown or `testdata.yaml`).

**Comment by niemela** [comment]
> > "example submissions" are the only kind of submissions in the problem packages.
> 
> yes but in the end the problem gets uploaded to a a system that handles both the example submissions and participants submissions. Now it has to distinguish them (which will likely also make it impossible to upload the example submission alone).

I disagree with that claim. A system does not *have* to distinguish submissions in the problem package after installation, from any user submissions. To upload package submissions that are using constants you would of course have to apply the constant replacements first, but that is exactly why the target system does not have to even know that our constants exist.

> Overall i am just in favor of not having constants in submissions and use the for their main purpose, to make checkers and statements consistent.

The consensus certainly seem to be that statements and validators are the more important use case. Is there anybody who sees a good use for constants in submissions (and included code)?



> [...] I would prefer to change the constants like this:
> 
>  * If its c++ do _X_ (for example macros)
>  * If its python do _Y_ (for example insert a variable with that value)
>  * If its tex do _Z_ (for example introduce a macro)
>  * Else, replace `{{constant}}` with its value (we don't (yer?) have better ideas for ctd checkers, viva checkers, markdown or `testdata.yaml`).

Ok, that would work, but:
- It's more complicated
- It still has the issue you are complaining about. The progams wont work until *after* you have made the replacements. If you're willing to do the replacements first (as is the intent) with these rules, why wont it work with the simpler rules?

**Comment by mzuenni** [comment]
> > > "example submissions" are the only kind of submissions in the problem packages.
> > 
> > 
> > yes but in the end the problem gets uploaded to a a system that handles both the example submissions and participants submissions. Now it has to distinguish them (which will likely also make it impossible to upload the example submission alone).
> 
> I disagree with that claim. A system does not _have_ to distinguish submissions in the problem package after installation, from any user submissions. To upload package submissions that are using constants you would of course have to apply the constant replacements first, but that is exactly why the target system does not have to even know that our constants exist.

Well, I think the system you upload the package to has to do the replacement (so for example domjudge)? 
Anyway, if I want to upload an example submission independently it won't work. As a problem setter, this makes it unlikely for me to use this feature for submissions. I get highlighting errors, I get syntax errors from my ide compiler, and I can't upload my submission and that is only for the small benefit of making changing one or two constants automatic? 

> Ok, that would work, but:
> 
> * It's more complicated
> * It still has the issue you are complaining about. The progams wont work until _after_ you have made the replacements. If you're willing to do the replacements first (as is the intent) with these rules, why wont it work with the simpler rules?

Yes, but I rarely run validators on my own (unlike submissions)? And with what I propose at least I don't get errors from my ide while writing those validators.
Further, for c++ my ide (and even plain g++ with `-D`) can do macro insertion. Same for latex (`-usepretex`). But none of my tools can directly do the replacement we currently specify. So I would need to write some script to add an extra build step that somehow does the replacement and in case of `latexmk -pvc` (continual rebuild) even that won't be sufficient...
Overall, this does not solve all problems especially if the language in question does not have a more convenient way of insertion but IMO in the cases where it is possible, we should use those features.

Also, yes what I propose makes it harder for us as tool authors but it makes it more convenient for the people who use our tools? In the end, our goal should be that constants are used as much as possible for the purpose of consistency i.e. ensuring that validators and therefore test cases match what is specified in the problem statement. Therefore, my aim is to make it as convenient as possible.

**Comment by niemela** [comment]
> Well, I think the system you upload the package to has to do the replacement (so for example domjudge)?

Yes, true, but only when installing the problem (or, depending on how you look at it, *before* installing them). And that point user submissions do not exist, so I would disagree with the statement that "it has to distinguish them".

> Also, yes what I propose makes it harder for us as tool authors but it makes it more convenient for the people who use our tools? In the end, our goal should be that constants are used as much as possible for the purpose of consistency i.e. ensuring that validators and therefore test cases match what is specified in the problem statement. Therefore, my aim is to make it as convenient as possible.

Good point.

**Comment by niemela** [comment]
I'm (again) hearing strong consensus for constants in statement and validators, some good arguments against constants in submissions, and (so far) no-one that explicitly wants constants in submissions (which is a lower bar than having a good argument for it), only that it could in theory maybe be useful (which is an even lower bar).

If you really think we should allow constants in submissions (and included code), please speak up.

**Comment by jsannemo** [comment]
@simonlindholm is included code hacks for faster interactive validators still a relevant thing? If so, this might be nice.

Also, just because constant replacement exists for user submissions, nothing forces you to use them if you find them inconvenient.

**Comment by simonlindholm** [comment]
> @simonlindholm is included code hacks for faster interactive validators still a relevant thing?

It is, and included code hacks are relevant for other weird kinds of multipass stuff as well (e.g. you can imagine two submission instances running in parallel and communicating with each other, which wouldn't be covered by the newly added multipass support, or you might want time limits to be across the sum of runs).

I personally wouldn't want to use `{{}}`-style constants for them though, it makes it too painful to test the code outside of verifyproblem. Instead I'd prefer to embed constants in the test data that the included code reads from stdin. (In case of fake interactive problems you also want to make them real interactive for judge UI purposes, so the interactive output validator would be responsible for communicating constants across at the start of the run, reducing the problem to that of getting output validators to know the constants.)

For input/output validators I think injected C++ defines could be decent enough UX, but for a more language-portable solution we could also use cmdline arguments (with some custom syntax for embedding constants in "*_validators_args" yaml strings?) or environment variables.

**Comment by eldering** [comment]
I don't have a strong opinion either way for allowing constants in submissions. And I agree with @niemela , this only apply to jury/example submissions, it's completely unrelated to team submissions, because any constant substitution should happen before these jury submissions are uploaded to the CCS (at least for DOMjudge), so I'd consider it part of the "exported" problem package, like calculated time limits and generated test data.

**Comment by evouga** [comment]
The part that I will definitely require my judges to use are constants in the validators, problem statement, and testdata.yaml.

I.e. in problem.yaml:
```
constants:
  minn: "1"
  maxn: "100"
  maxcoord:     "1e9"
  maxcoord_tex: "10^9"
  failstring: "No Solution"
  tol:     "1e-8"
  tol_tex: "10^{-8}"
```

In problem.en.tex:
```
The first line of input contains one integer $n ({{minn}} \leq n \leq {{maxn}})$, the number of points. 
The following $n$ lines will contain two integers each, $x$ and $y$, satisfying $0 \leq x,y \leq {{maxcoord_tex}}.$

Print the radius of the smallest disk containing all $n$ input points, if such a circle exists. 
Otherwise print \texttt{{{failstring}}}. 
Your answer will be judged correct if it within relative or absolute error ${{tol_tex}}$.
```

In validator.ctd:
```
INT({{minn}}, {{maxn}}, n) NEWLINE
REP(n)
INT(1, {{maxcoord}}) SPACE INT(1, {{maxcoord}}) NEWLINE
EOF
```

In testdata.yaml:
```
output_validator_args: "float_tolerance {{tol}}"
```

In a potential output validator:
```
std::string s;
std::getline(std::cin, s);
if(s == "{{failstring}}") {
  // ...
}
```

1. As was mentioned at the CLI, it is annoying that I need to use separate _tex constants, but I am OK with this as there is not any great alternative. The above is already much better than the status quo, as scanning the constants in one place is less error-prone than across multiple files, especially when many people are making changes to bounds to tweak time limit behavior etc.

2. I (still) think that constants in submissions are potentially useful, i.e.:
```
int main()
{
    for (int i = 0; i < {{maxn}}; i++)
    {
        // some math
        if (ok)
            std::cout << std::setprecision(30) << answer << std::endl;
        else
            std::cout << "{{failstring}}" << std::endl;
    }
}
```
(which does not "badly" mess up anything in my editor at least (VS); though the editor does try to insert a space between the "{{"). Inconsistencies in i.e. capitalization of `{{failstring}}` arise not infrequently during problem development. Mzuenni and others above have convinced my I probably won't be able to force my judges to use them, since independent compilation is indeed useful. But *I* would use them in my solutions, and I also don't see the harm in allowing them, unless there's a risk of accidental constant replacement in code that's not intended to have constants?

3. I agree with @niemela that independent compilation of the .tex, validators, etc. is fundamentally at tension with the whole point of the global constants (which is to tightly couple these components in a way that reduces the incidence of broken problems). If people *really* want to independently compile the .tex etc. and it is possible with a few light scripts (especially if scripts to do the constant replacement are available in the tools and easy to use), I think that is OK.

4. I am OK with some kind of language-specific idiomatic injection scheme, as long as it Just Works for all constants I want to use in practice. That includes all types of literals (including floating-point numbers and big integers) and strings (including with internal spaces as in the above.) (Pasting boilerplate into all of my validators and problem statements to parse these things from environment variables does not count as "Just Works.") I am still not entirely convinced that {{this}} syntax is *all that bad*, though. The above snippets were easy to write and it is clear at a glance where the global constants are.

**Comment by Tagl** [comment]
> I am still not entirely convinced that {{this}} syntax is all that bad, 
What about conflicts with, for example, Python's f-strings?

**Comment by eldering** [comment]
> > I am still not entirely convinced that {{this}} syntax is all that bad,
> 
> What about conflicts with, for example, Python's f-strings?

Not really though, right, because those only use single `{this}`, and also, you can always declare a variable and then use that

**Comment by niemela** [comment]
> 1. As was mentioned at the CLI, it is annoying that I need to use separate _tex constants, but I am OK with this as there is not any great alternative. The above is already much better than the status quo, as scanning the constants in one place is less error-prone than across multiple files, especially when many people are making changes to bounds to tweak time limit behavior etc.

Another solution would be write some formatting function in tex, so that instead if using `{{tol_tex}}` you would use `\format_nicely{{{tol}}}`. I would think that due to the simplicity of the `_tex` constants, they are more likely to be used, and the downside (i.e. that `{{tol}}` and `{{tol_tex}}` could actuall in theory become out of sync) is mitigated by the fact that they will be defiend *right* next to each other.

**Comment by simonlindholm** [comment]
> (Pasting boilerplate into all of my validators and problem statements to parse these things from environment variables does not count as "Just Works.")

FWIW for input validators I somewhat disagree -- CTD can use the custom syntax, and any non-CTD syntax validator will need to use some kind of testlib/template to correctly validate whitespace/no trailing zeroes/etc. because it's too error-prone to hand-roll. If you combine CTD and non-CTD you're probably putting the limits in the CTD file, but even if not it's really not hard to read data from env vars: in Python it's `int(os.getenv("MYVAR"))` and in C++ `atoi(getenv("MYVAR"))`.

**Comment by RagnarGrootKoerkamp** [comment]
- I'm neutral on substitution in jury submissions. Don't mind supporting it but not the critical use case.
- For latex, introducing `\maxn` and the like sounds good.
- For Python and C++ validators, initially I'm not a big fan of environment variables, but it may come out quite clean. One thing that could then be done is that what is currently `v.read_int("a", 0, 1000)` could then just become `v.read_int("a")`, which can read the `mina` and `maxa` environment variables automatically. So actually this would lead to *less* code :)
- Question then is maybe is we should 'standardize' that such bounds are *inclusive*. It's still easy to mess up `a \leq \maxa` and `a < \maxa`. But I think exclusive bounds are quite rare, so maybe a non-issue.
- BAPCtools specific, but we should also pass these to generators in the same way as we do for validators. It would be nice if command line args `maxa=100` can in some way cleanly override equivalent environment variables without needing to include a library.

**Comment by evouga** [comment]
>  any non-CTD syntax validator will need to use some kind of testlib/template to correctly validate whitespace/no trailing zeroes/etc.

This is true for output validators, but not always for input validators: one usage pattern is to have CTD handle the input syntax and for the C++/python validators to check more complicated semantics assuming correct syntax.

Anyway, I have a negative bias against environment variables, but maybe it is clean? What would my example above look like in this scheme? LaTeX would also use environment variables, or something else? It would work when the variables have TeX embedded in them?

**Comment by evouga** [comment]
```
> One thing that could then be done is that what is currently v.read_int("a", 0, 1000) could then just become v.read_int("a"), which can read the mina and maxa environment variables automatically. So actually this would lead to less code :)
> Question then is maybe is we should 'standardize' that such bounds are inclusive. It's still easy to mess up a \leq \maxa and a < \maxa. But I think exclusive bounds are quite rare, so maybe a non-issue.
```
My sense is that this is too narrow/special-case. Some of the North American judges have *very* strong opinions about the style of the problem statement...

**Comment by niemela** [comment]
There does not seem to be a clear consensus happening. My reading, but I could be biased because it's also my opinion, is that no changes are needed.

I will close this shortly if there are no further comments.

**Comment by eldering** [comment]
I think it's worth to try to come to a consensus of a best solution here, as changing it later would be hard.

Some thoughts:

- We could in principle do both `{{name}}` substitution *and* passing constants via environment variables. That would give the problem author a choice. Not sure if this is a good thing to do though...
- Although constants can be passed via environment variables to (jury) submission by problem/bapctools easily, if we upload those submissions e.g. to DOMjudge, then that is more annoying than if the constants were just substituted into the code.

**Comment by niemela** [comment]
> We could in principle do both `{{name}}` substitution _and_ passing constants via environment variables. That would give the problem author a choice. Not sure if this is a good thing to do though...

IMO, most of the arguments against the current state would be worse (or similar) if we do both. I don't like "both".

> Although constants can be passed via environment variables to (jury) submission by problem/bapctools easily, if we upload those submissions e.g. to DOMjudge, then that is more annoying than if the constants were just substituted into the code.

Agreed.

This still sounds like status quo is the best option?

**Comment by RagnarGrootKoerkamp** [comment]
Do you mean by 'current state' either a) not having constants, or b) the `XXXconstantXXX` or so that IIRC we merged at some point?

I'm kinda torn on this, but I actually think that I slightly prefer doing both substitution and env-vars. Then validator libraries can read the env, while submissions don't have to bother with that and can just use the constant. For latex specifically, it makes most sense to define latex macros directly IMO.

@mzuenni what's your current thinking here?

**Comment by niemela** [comment]
> Do you mean by 'current state' either a) not having constants, or b) the `XXXconstantXXX` or so that IIRC we merged at some point?

No, the 'current state' is the `{{name}}` constants that are currently described here: https://www.kattis.com/problem-package-format/spec/2023-07-draft.html#constants .

**Comment by mzuenni** [comment]
For latex I would say the best option is to inject them in a latex like fashion, i.e, use `-usepretex`.
Then we could either inject individual commands or a `\constants{name}` command. I think the later one is preferable.

For validators, I definitely prefer the solution with environment variables, since this still allows compiling them on the console and parsing can be done by headers.

For submissions, I am less convinced, but being able to compile and run them is pretty useful...
I played around with the substitution and ended up writing stuff like `VAR = parse("{{constant}}", default)` which is kind of annoying...

**Comment by niemela** [comment]
> I played around with the substitution and ended up writing stuff like `VAR = parse("{{constant}}", default)` which is kind of annoying...

I don't understand why substitution leads to this? Could you elaborate/explain?

**Comment by mzuenni** [comment]
>>I played around with the substitution and ended up writing stuff like VAR = parse("{{constant}}", default) which is kind of annoying...

>I don't understand why substitution leads to this? Could you elaborate/explain?

Well I wanted to have a program that I could compile on my own and that was the easiest way to do that ^^'

**Comment by RagnarGrootKoerkamp** [comment]
We could also decide to just not do substitution for submissions. I agree that the value there is limited, since it's very unlike that all submissions to a problem will use the problem defined placeholders, and so if bounds change, probably either none or most submissions need updating anyway.

Then that leaves the latex, where I agree that `\constant{name}` is nice (I'd make it singular). This then also allows `\contant{n_max}`, whereas `\consant_n_max` is broken.

For the validators, a library is usually used anyway.
A benefit of environment variables then is that they can be read dynamically: A user can write `read_int("n")` and the library can automatically translate that into reading environment variables `n_min` and `n_max`. With substitution, that would be harder.

**Comment by RagnarGrootKoerkamp** [comment]
Ah did we settle on how latex constants should be formatted?
If we go with `\constant{name}`, we could avoid doing the formatting in latex itself, and provide `\constant{name-<variant>}`, where we provide a few different pre-formatted variants.

**Comment by evouga** [comment]
Does checktestdata support environment variable constants? The spec doesn't mention them: https://github.com/DOMjudge/checktestdata/blob/main/doc/format-spec.md

I include a CTD syntax validator for pretty much every problem I write and I definitely want constants in there.

**Comment by RagnarGrootKoerkamp** [comment]
Yeah we should do dedicated Syntax for .ctd then. Want to propose something? (I haven't used it in a long time.)

**Comment by mpsijm** [comment]
For problem development, I think I prefer the env-var approach for everything (besides LaTeX, where defining commands works better). But how will this work when exporting the problem to e.g. DOMjudge, will that system then have to pass the constants to e.g. the output validator/visualizer? Or should the development tooling perform hacky substitution anyway when exporting a ZIP? @eldering you mentioned this would be "more annoying for submissions", but the output validator/visualizer are examples of other dynamic parts of the problem that the contest system should support 🤔 Probably, we want all constants inlined once we're at the point that the ZIP gets exported to some archive, unless we're certain that anybody who would read those ZIPs support the latest version of the spec.

Whether the env-vars should be readable by submissions at all, I'm not entirely sure about 😛 I think it'd be nice to have during development, but indeed, it does make things more complex to debug. I saw it mentioned that you could write a submission without constants during development and only rewrite using constants after, but I can see the need to go back-and-forth between those two states.

Something completely different for which I'm probably a bit late to the party: the idea of defining constants at a higher level is nice, but there will always be scenarios in which the constants will not suffice. For instance, when generating tests (I'm thinking of BAPCtools' `generators.yaml` here), you may want to generate additional tests when `n_max` is raised from 1000 to 10000, rather than only changing the maximal cases. Potentially, adding constants will make it easier to forget these kinds of things, especially for people who are new to the jury process.

**Comment by niemela** [comment]
> For problem development, I think I prefer the env-var approach for everything

Could someone explain to me more exactly how the "env-var appoach" would work?

> But how will this work when exporting the problem to e.g. DOMjudge, will that system then have to pass the constants to e.g. the output validator/visualizer? 

I don't understand how that would *not* be the case? (but, see my question above).

> Or should the development tooling perform hacky substitution anyway when exporting a ZIP? 

I think that would be quite terrible.

> @eldering you mentioned this would be "more annoying for submissions", but the output validator/visualizer are examples of other dynamic parts of the problem that the contest system should support 🤔 Probably, we want all constants inlined once we're at the point that the ZIP gets exported to some archive, unless we're certain that anybody who would read those ZIPs support the latest version of the spec.

I think we should assume that systems that use the format understands the format.

**Comment by mpsijm** [comment]
@mzuenni drafted an implementation of this at https://github.com/RagnarGrootKoerkamp/BAPCtools/pull/430. In short, the constants are added to the environment variables of the program that is about to be run: https://github.com/RagnarGrootKoerkamp/BAPCtools/pull/430/files#diff-d9a2c5209d9ce106e26d6562de2c7ac413b2d0819a9dbd018f75d7056f8e8ffdR511 (the linked `kwargs` are later passed to a subclass of `subprocess.Popen` in `util.exec_command`)

**Comment by mzuenni** [comment]
> I think we should assume that systems that use the format understands the format.

But do we want to assume that they use the full spec?  For example, right now DOMjudge does not use the time multipliers to determine the time limit and will not compile the tex code.

>>But how will this work when exporting the problem to e.g. DOMjudge, will that system then have to pass the constants to e.g. the output validator/visualizer?
>
>I don't understand how that would not be the case? (but, see my question above).

If we use the Constant sequences as they are defined in the draft right now we could replace them with the constant value while exporting the package. This would make it possible that the system where the packages are imported would not need to know about constants.

After playing around with both the `env-var` method and the text replacement method, my opinion is the following:
- For Latex we should provide the command `\constant{<key>}`. We already provide commands like `\nextSample` etc. so this is the most natural extension.
- Everywhere else we should just use the text replacement of `{{key}}` as it is already described in the draft. My reason here is that we still have a few different file types (`.cpp`, `.py`, `.ctd`, `.viva` and `.yaml`) and the `env-var` method does not work for all of them. And the constant replacement has the benefit of being more backwards compatible.

**Comment by eldering** [comment]
> Does checktestdata support environment variable constants? The spec doesn't mention them: https://github.com/DOMjudge/checktestdata/blob/main/doc/format-spec.md
> 
> I include a CTD syntax validator for pretty much every problem I write and I definitely want constants in there.

The checktestdata program (not the specification) has the option `--preset=VAR=VALUE,...` that can be used to pass constants in (and override any values set in the CTD file).

**Comment by eldering** [comment]
> > For problem development, I think I prefer the env-var approach for everything
> 
> Could someone explain to me more exactly how the "env-var appoach" would work?

My understanding is that when a program like a jury submission, or input/output validator is called, the problem constants are declared as environment variables and then the program can grab them from the environment.

> > But how will this work when exporting the problem to e.g. DOMjudge, will that system then have to pass the constants to e.g. the output validator/visualizer?
> 
> I don't understand how that would _not_ be the case? (but, see my question above).
> 
> > Or should the development tooling perform hacky substitution anyway when exporting a ZIP?
> 
> I think that would be quite terrible.
> 
> > [@eldering](https://github.com/eldering) you mentioned this would be "more annoying for submissions", but the output validator/visualizer are examples of other dynamic parts of the problem that the contest system should support 🤔 Probably, we want all constants inlined once we're at the point that the ZIP gets exported to some archive, unless we're certain that anybody who would read those ZIPs support the latest version of the spec.
> 
> I think we should assume that systems that use the format understands the format.

My biggest issue with passing environment variables is for jury submissions. We currently treat them basically as normal submissions (just submitted by a team in jury/admin group) and judge them the same as team submissions. If we'd have to pass problem constants as environment variables, then that would also be exposed to team submissions, which I think is a security hole. OTOH, I think that treating jury submissions different from team submissions is not good either because then we don't properly test these submissions in the same way anymore.

For validator and visualizer scripts, this is not really an issue, but I'd still say it would be a bit inconvenient to have to separately import the problem constants and pass them as environment variables to these scripts. So in general I'd much prefer getting/creating from bapc/problemtools an "exported" or "frozen" version of a problem package where these constants are hardcoded in the programs.

**Comment by niemela** [comment]
> > I think we should assume that systems that use the format understands the format.
> 
> But do we want to assume that they use the full spec? For example, right now DOMjudge does not use the time multipliers to determine the time limit and will not compile the tex code.

I feel that I don't really understand your point. We *don't* assume that any system/user uses the full spec. The spec is very modular in the sense that you can just *not* use parts of it, such as constants, scoring, multi-pass, and many other. But if you **do** use some part, then you understand that part, and if you don't understand some part, then you don't support using that part.

> If we use the Constant sequences as they are defined in the draft right now we could replace them with the constant value while exporting the package. This would make it possible that the system where the packages are imported would not need to know about constants.

Right, and that's (IMO) good. (And maybe that's your point above?)

> After playing around with both the `env-var` method and the text replacement method, my opinion is the following:
> 
>   * For Latex we should provide the command `\constant{<key>}`. We already provide commands like `\nextSample` etc. so this is the most natural extension.

I'm not *sure* it's worth the extra complexity... but it's not a *lot* of extra complexity, and it is a little bit nicer...

>   * Everywhere else we should just use the text replacement of `{{key}}` as it is already described in the draft. My reason here is that we still have a few different file types (`.cpp`, `.py`, `.ctd`, `.viva` and `.yaml`) and the `env-var` method does not work for all of them. And the constant replacement has the benefit of being more backwards compatible.

That seems good to me. It's definitely simpler, and the person that would suffer from the con of this (i.e. the problem setter) is the one that decides whether to use it or not.

**Comment by niemela** [comment]
> My understanding is that when a program like a jury submission, or input/output validator is called, the problem constants are declared as environment variables and then the program can grab them from the environment.

having example submissions be written to read data form environment variables seems **very** bad to me. That will not work as a normal submission, and the translation needed is not trivial. Search and replace, OTOH, **is** trivial.

> For validator and visualizer scripts, this is not really an issue, but I'd still say it would be a bit inconvenient to have to separately import the problem constants and pass them as environment variables to these scripts. So in general I'd much prefer getting/creating from bapc/problemtools an "exported" or "frozen" version of a problem package where these constants are hardcoded in the programs.

The "exported" or "frozen" form is very reasonable, but really an internal implementation detail.

**Comment by RagnarGrootKoerkamp** [comment]
Ok, we agree on `\constant{name}` in latex and `{{name}}` in validators then.

I didn't see any further opinions on whether we should support it in submissions. I'm pretty much neutral. Nice to have in the ocasional case where it's useful, but I expect most submissions to anyway not use the constants.

Re the 'frozen' package: Yeah, so from the BAPCtools perspective we're probably not expecting DOMjudge to implement this for now, so we can just run the substitution on export alongside the other fixups we're already doing anyway.
This feature really is mostly for development anyway, and there's not really much benefit once the package is done.

For BAPCtools, we can also do substitution in the generators then.

**Comment by mpsijm** [comment]
If the constants are only for problem development and we're going for substituting `{{key}}`, I guess we can support it in submissions as well? Most of the discussion regarding constants in submissions arose from having to treat them differently, but that's not a concern if we're providing a "frozen" exported ZIP anyway. And if the problem setter doesn't like using constants in jury submissions, then they simply don't 😄 

> I'm not _sure_ [`\constant{key}` is] worth the extra complexity... but it's not a _lot_ of extra complexity, and it is a little bit nicer...

Slight agree on this btw: if we use `{{key}}` in all other files, we might as well use that in LaTeX as well to keep things consistent. That also keeps exporting a frozen package slightly simpler, but again, not by much, so if others prefer `\constant{key}` in LaTeX, I'm fine with it 🙂

**Comment by niemela** [comment]
I think we agreed that #370 is what we wanted to change, and the rest should stay as is. Closing this as done.

If I missed something feel free to reopen... BUT, I think it would be better to create a new issue with only what is still remaining.

**Comment by eldering** [comment]
I'm reopening this issue instead of creating a new one since I think it makes more sense to keep the full discussion history in one place and I'm addressing the whole issue, not a specific extension/detail.

So the trigger of this is that at BAPC 2025 a change was made to the limit of one of the problems that ended up in the test data and validator, but not the problem text. We then had a discussion of how we could prevent this best with @RagnarGrootKoerkamp , @mpsijm and me mainly.

So an alternative approach to trying to substitute the constants in all places, would be to annotate them, so tooling can check that the values in all places matches. The advantage would be that no files have to be rewritten (which I think is currently not well-specified when/how this happens: does the file get overwritten, does happen on "export" or when running it only, ...), as well as issues with the `{% ... %}` template conflicting with language syntax. The difficulty would be to find a way to annotate things across languages so that each constant and its value can be consistently detected.

For languages that have comments like C++ that can start and stop anywhere that would be easy with a syntax like
```
int max_n = /*@@constant:max_n@@ */ 1000 /*@@constant_end@@ */ ;
```
or something similar.

For languages that only support until end of line comments and where newlines have syntactical meaning, like python and shell script, this might be a bit more difficult. Maybe some heuristic could still be used, like:
```python
# @@constant_on_next_line@@
max_n = 1000
```
Assuming that all constants are declared, it would not be very hard to detect the constant and its value here. And the tool could complain if it can't correctly detect it.

**Comment by mzuenni** [comment]
> The advantage would be that no files have to be rewritten (which I think is currently not well-specified when/how this happens: does the file get overwritten, does happen on "export" or when running it only, ...)

BAPCtools does this on export, this ensures that the judging system does not need to know about constants in the same way it does not need to know about input valdiators

> The difficulty would be to find a way to annotate things across languages so that each constant and its value can be consistently detected.

The same could have been done for constants... and we decided to not do this ^^'

Overall i think the idea is interesting but I also think its way more difficult then what we have right now. And there is the additional issue that the constant value in different programs will look different (e.g. `10^5` in latex but `100'000` in c++), so automatically checking them is not really feasible either (that issue is true for constants as well but at least the value of the constant is displayed at the same point in the `problem.yaml` which makes manual comparison easier).

**Comment by niemela** [comment]
> So the trigger of this is that at BAPC 2025 a change was made to the limit of one of the problems that ended up in the test data and validator, but not the problem text.

So *exactly* the thing constants are intended to fix.

> So an alternative approach to trying to substitute the constants in all places, would be to annotate them, so tooling can check that the values in all places matches.

Why is an alternate method needed? Did you use constants? If you didn't, wouldn't that have solved the problem?

> BAPCtools does this on export, this ensures that the judging system does not need to know about constants in the same way it does not need to know about input valdiators

Does this mean that problem packages created by BAPCtools never use constants (since they are effectively removed when "exporting").

> Overall i think the idea is interesting but I also think its way more difficult then what we have right now.

Strongly agree.

> By the way: tooling could warn if a constant is defined in `problem.yaml` but not used in a solution or input validator. This _does_ seem useful and will catch judges hard-coding the constants instead of using the problem-defined ones.

That's a good idea. (One would need to remember that "used in an input validator" should include both being in the validator code, or given as an argument in `test_group.yaml` though).

> EDIT: although on second thought, the need for multiple constants for different formatting in .tex etc. prevents this, unless there is a way in the `.yaml` to say that two constants are "the same".

There are certainly heuristics one could use.

**Comment by niemela** [comment]
I have a strong feeling that this could and *should* be closed again. I will likely do so quite soon unless someone argues against that.

**Comment by Matistjati** [comment]
> By the way: tooling could warn if a constant is defined in problem.yaml but not used in a solution or input validator. This does seem useful and will catch judges hard-coding the constants instead of using the problem-defined ones.

At the very least, we should definitely check that every constant is used at least once in the problem package.

**Comment by thorehusfeldt** [comment]
Aside about `\constant{name}` in LaTeX: 

What we seem to want to support is a mechanism for the Latex source to refer to values defined in `problem.yaml`. I clean way to do that (which would require minimal effort do define) is `\problem{foo}` where `foo` is a key in `problem.yaml`. 

Then you could write `\problem{name}` (to refer to the problem title) and `\problem{constants.maxn}` or `\problem{constants.maxn.se.tex}`. Conceptually, these are the same thing. (How much syntactic sugar should be heaped on these macros is a different question.)

This is similar to providing all python code with a namespace `problem`. Then we can have

```python
maxn: int = problem.constants.maxn
no_answer: string = problem.contants.no_answer
```

There may be a similarly well-established way for `generators.yaml` to get access to the named yaml file “one level up.” (Honestly I don’t know.)

Ideally, there would be clean ways to offer a well-defined subset of programming languages a well.defined ways to access `problem.yaml`.

**Comment by RagnarGrootKoerkamp** [comment]
After thinking about it some more and trying to implement it in BAPCtools, we now think that substituting constants in jury-provided submissions is not ideal and we'd prefer to revert that part (but keep them in validators and the statement).

<whoops hit ctrl-enter too soon; still editing>

**Comment by mpsijm** [comment]
Note that my comment is from before I realized that (to quote Ragnar) "mixing frozen submissions and unfrozen submissions" can be problematic.

To add some more argumentation why `submissions/` don't strictly need constants, partly summarized from our private chat:
- Most jury submissions don't need to explicitly mention the input bound constants, since they read the input from stdin.
- If they do, tooling (like `bt run` or `verifyproblem`) will flag these submissions as faulty when the constants change anyway.

The most important thing is that the constants in the statement match with the test data (by input validators/generators).

**Comment by niemela** [comment]
So, what I think I'm hearing is that we should drop "example submissions" from the list of things that constants are replaced in?

I would argue that logically, we should then also drop "included files". Do you agree?

What about `test_group.yaml`?

**Comment by Matistjati** [comment]
> What about test_group.yaml?

I'm strongly opposed to dropping this. If we drop it, we lose the ability to sync the value of test groups and the reported text in the statement.

Much weaker, but in Sweden, we make heavy use of `input_validator_args` to validate subtask constraints. Some judges prefer to only use `input_validator_args`, and never hardcode constant in the input validator. Of course, this is just preference.

**Comment by Matistjati** [comment]
Further, I would argue that there are cases of "included files" where it constants would be very useful. As a concrete example, suppose you want to make your problem interactive, but it's too slow to use the normal interactive because of context switches. Then, you can provide "runners" in multiple languages. We can't really have data secret from the submission because the runner won't be able to read the `.ans` file, and submissions can quite easily find the `.in` file. In this case, constants would allow you to synchronize constants cross runners from different languages.

Even though this is a niche example, the problem format has always had a history of people implementing unsupported problem types (multi-pass, static validators, forced online without massive overhead) using included files. I do hope that in a couple of years, people create new interesting things using included files, in which case constant substitution makes it a little bit easier. 

I feel like this is a compelling point to keep constant substitution for included files, and haven't seen any concrete arguments against it.
